### ISSUE TITLE
md4all-main/models/md2/depth_decoder.py

### ISSUE THREAD

**Issue by zsh1223** [md4all/md4all#25]
In depth decoder. py, I found that during initialization, you set scale=range (4) and implemented 
"for s in self. scales:
Self. convs [("dispconv", s)]=Conv3x3 (self. num_ch_dec [s], self. num_output_channels)"
But in def forward(), i loops 5 times. Why is this?
Scales contain four scales, namely 0, 1, 2, and 3. So in def forward(), when i=4. Does this mean that the decoder is restoring the original image size?

**Comment by morbi25** [comment]
Hi @zsh1223,

Thank you for your question! :)

As I understand, you are wondering why the loop in the forward function in the depth_decoder.py file has five iterations (4, 3, 2, 1, 0) while having only four scales (0, 1, 2, 3). To make that more clear, it is beneficial to look at the forward function of the depth_decoder.py file that we reused from the [monodepth2 code repository](https://github.com/nianticlabs/monodepth2/blob/master/networks/depth_decoder.py):
```python 
def forward(self, input_features):
    self.outputs = {}

    # decoder
    x = input_features[-1]
    for i in range(4, -1, -1):
        x = self.convs[("upconv", i, 0)](x)
        x = [upsample(x)]
        if self.use_skips and i > 0:
            x += [input_features[i - 1]]
        x = torch.cat(x, 1)
        x = self.convs[("upconv", i, 1)](x)
        if i in self.scales:
            self.outputs[("disp", 0, i)] = self.sigmoid(self.convs[("dispconv", i)](x))

    return self.outputs
```
This function propagates the encoded input through the five layers of the depth decoder network specified in Table 5 of the supplementary material of the [monodepth2 paper](https://arxiv.org/pdf/1806.01260.pdf). Therefore, the loop has five iterations.
Furthermore, monodepth2 utilizes (full-resolution) multi-scale estimation to prevent the training objective from getting stuck in local minima. This means that the depth decoder outputs estimates at different resolutions. In our case, we have four scales such that the decoder network produces estimates at four different resolutions (i.e., original image resolution downscaled by factor 2^3, 2^2, 2^1, 2^0), each coming from the respective intermediate layer. For example, for the image resolution (576, 320, 3) that we used for nuScenes, the depth decoder outputs estimates with resolution (72, 40, 1) from layer 4, (144, 80, 1) from layer 3, (288, 160, 1) from layer 2, and (576, 320, 1) from layer 1.
Please note that similar to md2, we employ **full-resolution** multi-scale estimates, i.e., at a later stage, we upsample the lower resolution estimates from the depth decoder's intermediate layers to match the original image resolution before performing the reprojection and computing the photometric error. For more details, please refer to the subsubsection 3.2 => Multi-scale Estimation of the [monodepth2 paper](https://arxiv.org/pdf/1806.01260.pdf).

I hope this answer helps you! Have a nice week. :)

Best regards,
Nils

**Comment by zsh1223** [comment]
Thank you for your answer.I have reproduced the baseline in your code. The "night" item in my reproduced results did not achieve the excellent results in your paper, and I only achieved an accuracy of 0.5375. Why is this?
![image](https://github.com/md4all/md4all/assets/130481211/f690684f-1547-4216-aa1b-11257c227a28)
Among them, baseline0 is the pre trained model you provided, and baseline1 is my own training from scratch.

**Comment by morbi25** [comment]
Hello @zsh1223,

Thank you for reaching out again!

Based on the current details you shared with us, pinpointing the exact issue is very challenging. Please ensure that you set up everything correctly, used the correct configurations for training, etc., as described in our README. We highly recommend using the docker environment to increase reproducibility. :)

Best regards,
Nils

**Comment by zsh1223** [comment]
Can your program use mini datasets
![image](https://github.com/md4all/md4all/assets/130481211/0d1b5665-5550-48c4-bb93-f7b2dab26ac5)
 i use the“_C.DATASET.VERSION = 'mini'” in config.py.The error is as follows:
![image](https://github.com/md4all/md4all/assets/130481211/a8c89d53-100e-48f0-86a4-6b2a2e319812)

**Comment by zsh1223** [comment]
I have another question：
What should I do if I want to test a single image using baseline.ckpt.I would greatly appreciate it if you could answer my question.

**Comment by morbi25** [comment]
Hello @zsh1223,

Thank you for your questions and sorry for the late reply!

1. Yes, the nuScenes mini dataset should be supported. Please make sure that your specified data root directory (configurable with DATASET.DATAROOT) contains the "v1.0-mini" folder (for meta information). From the console output, it seems to be missing.
2. If you want to test a single image with our models, please use the "test_simple.py" file. For more information, see #2.

I hope these answers help you! :)

For future questions, please also consider opening new issues if your questions address new domains. Thank you in advance! :)

Best regards,
Nils

**Comment by zsh1223** [comment]
![image](https://github.com/md4all/md4all/assets/130481211/b2fb9133-d5aa-4cd4-b2ec-fcb0f279a2aa)
1.I have set up my directory and config, as shown in the picture.
2.I know your model can be tested using "test_simple. py" for image testing. But this was tested using the DDa model, which is a knowledge distillation model. What I want is to only use the baseline model (student model).

**Comment by morbi25** [comment]
Hi @zsh1223,

1. As shown in the "Datasets" section of our repository [here](https://github.com/md4all/md4all?tab=readme-ov-file#datasets:~:text=r%20requirements_w_version.txt-,Datasets,-%3A), our code assumes a subfolder that separates the different nuScenes splits, i.e., in your nuscenes folder you need to add another folder for "trainval", "test", and also for "mini". These subfolders should then contain the respective directories "maps", "samples", "sweeps", "translated" (optional), and the meta information (e.g., "v1.0-mini"). Please make sure that your dataset is structured accordingly as shown in the attached image:
![nuScenes_structure](https://github.com/md4all/md4all/assets/45937132/7f191d88-67de-4429-af1e-af71d50837ec)
2. The "test_simple.py" can also be used to test our baseline model on custom images. For that, you have to adapt the "LOAD.CHECKPOINT_PATH" parameter and remove the configuration of the whole "LOSS" configuration section. For simplicity, I attached an example configuration file to this message: [test_simple_baseline_nuscenes.txt](https://github.com/md4all/md4all/files/14742604/test_simple_baseline_nuscenes.txt)
As ".yaml" files are apparently not supported in a GitHub comment, I attached a ".txt" file, so please make sure to change the file type to ".yaml" in case you download and plan to use it. :)

**Comment by zsh1223** [comment]
Thank you very much for your answer. Your help is of great significance to me.
